File: java/queue/src/main/java/com/js/datastructuresalgorithms/queue/JSQueue.java
package com.js.datastructuresalgorithms.queue;

import java.util.LinkedList;

public class JSQueue<T> {

    //*****************
    // Class Members **
    //*****************
    private LinkedList<T> mList;

    //***************
    // Constructor **
    //***************
    public JSQueue() {}

    public void enqueue(T item) { list().addLast(item); }
    
    public T dequeue() { return list().removeFirst(); }

    public T peek() { return list().peekFirst(); }

    public int size() { return list().size(); }

    public boolean isEmpty() { return list().isEmpty(); }

    //*******************
    // Private Members **
    //*******************

    private LinkedList<T> list() { return mList == null ? mList = new LinkedList<>() : mList; }
}

File: node/queue/src/types/JSQueue.js
class JSQueue {

    #mArray = [];

    size = () => this.#array().length;

    isEmpty = () => this.size() === 0;

    enqueue = item => this.#array().push(item);

    dequeue = () => this.#array().shift();

    peek = () => this.#array()[0];

    //*******************
    // Private Members **
    //*******************

    #array = () => this.#mArray;
}

module.exports = JSQueue;

